cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java

```diff
@@ -0,0 +1,179 @@
+            id.getNamespace(), id.getName()));
+      }
+      writeIdentity(context, id, identity);
+    }, AlreadyExistsException.class, IOException.class);
+  }
+
+  public void update(CredentialIdentityId id, CredentialIdentity identity)
+      throws IOException, NotFoundException {
+    Utils.validateResourceName(id.getName());
+    TransactionRunners.run(transactionRunner, context -> {
+      if (!fetchRow(context, id).isPresent()) {
+        throw new NotFoundException(String.format("Credential identity '%s:%s' not found",
+            id.getNamespace(), id.getName()));
+      }
+      writeIdentity(context, id, identity);
+    }, IOException.class, NotFoundException.class);
+  }
+
+  public void delete(CredentialIdentityId id) throws IOException, NotFoundException {
+    Utils.validateResourceName(id.getName());
+    TransactionRunners.run(transactionRunner, context -> {
+      if (!fetchRow(context, id).isPresent()) {
+        throw new NotFoundException(String.format("Credential identity '%s:%s' not found",
+            id.getNamespace(), id.getName()));
+      }
+      StructuredTable table = context.getTable(CredentialProvisionerStore.CREDENTIAL_IDENTITIES);
+      Collection<Field<?>> key = Collections.unmodifiableList(Arrays.asList(
+          Fields.stringField(CredentialProvisionerStore.NAMESPACE_FIELD,
+              id.getNamespace()),
+          Fields.stringField(CredentialProvisionerStore.IDENTITY_NAME_FIELD,
+              id.getName())));
+      table.delete(key);
+    }, IOException.class, NotFoundException.class);
+  }
+
+  private Optional<StructuredRow> fetchRow(StructuredTableContext context,
+      CredentialIdentityId id) throws IOException {
+    StructuredTable table = context.getTable(CredentialProvisionerStore.CREDENTIAL_IDENTITIES);
+    Collection<Field<?>> key = Collections.unmodifiableList(Arrays.asList(
+        Fields.stringField(CredentialProvisionerStore.NAMESPACE_FIELD,
+            id.getNamespace()),
+        Fields.stringField(CredentialProvisionerStore.IDENTITY_NAME_FIELD,
+            id.getName())));
+    return table.read(key);
+  }
+
+  private void writeIdentity(StructuredTableContext context, CredentialIdentityId id,
+      CredentialIdentity identity) throws IOException {
+    // Validate the referenced profile exists.
+    StructuredTable profileTable = context
+        .getTable(CredentialProvisionerStore.CREDENTIAL_PROFILES);
+    Collection<Field<?>> profileKey = new ArrayList<>();
+    CredentialProfileId profileRef = identity.getCredentialProfile();
+    profileKey.add(Fields.stringField(CredentialProvisionerStore.NAMESPACE_FIELD,
+        profileRef.getNamespace()));
+    profileKey.add(Fields.stringField(CredentialProvisionerStore.PROFILE_NAME_FIELD,
+        profileRef.getName()));
+    if (!profileTable.read(profileKey).isPresent()) {
+      throw new IllegalArgumentException(String.format("Credential profile '%s:%s' not found",
```

Comment:
it might be good to throw a more specific exception here, especially if we want this to eventually result in a 400 (I think IllegalArgument gets translated into a 500?)

Reply:
Done, changed this to BadRequestException